---
title: "Bisque_POC_Using_Mouse_Kidney"
author: "Harry"
date: "1/4/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
library(tidyverse)
library(Biobase)
library(BisqueRNA)
library(data.table)
```

# Read in raw data from GEO

```{r dataread, include=FALSE}
# Read in singe cell RNA data
path_to_scRNA_data <- "C:\\Users\\hsmit\\Documents\\Saba_Lab\\R24\\Bisque_R24\\rawdata\\GSE107585_series_matrix\\"

file_name_scRNA_data <- "Mouse_kidney_single_cell_datamatrix.txt"

scRNA_raw_data <- fread(paste0(path_to_scRNA_data, file_name_scRNA_data))


# read in bulk RNA data
path_to_bulkRNA_data <- "C:\\Users\\hsmit\\Documents\\Saba_Lab\\R24\\Bisque_R24\\rawdata\\GSE81492_series_matrix\\"

bulk_file_names_list <- list.files(path = path_to_bulkRNA_data)

GSM2154800_G1NF_42 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[2]))
colnames(GSM2154800_G1NF_42)[c(1,2)] <- c("Gene_ID", "G1NF_42")
GSM2154801_GINF_44 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[3]))
colnames(GSM2154801_GINF_44)[c(1,2)] <- c("Gene_ID", "GINF_44")
GSM2154802_GNA_73 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[4]))
colnames(GSM2154802_GNA_73)[c(1,2)] <- c("Gene_ID", "GNA_73")
GSM2154803_GNA_78M <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[5]))
colnames(GSM2154803_GNA_78M)[c(1,2)] <- c("Gene_ID", "GNA_78M")
GSM2154804_G1NF_28 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[6]))
colnames(GSM2154804_G1NF_28)[c(1,2)] <- c("Gene_ID", "G1NF_28")
GSM2154805_GNA1_18 <-fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[7]))
colnames(GSM2154805_GNA1_18)[c(1,2)] <- c("Gene_ID", "GNA1_18")
GSM2154806_NA_27 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[8]))
colnames(GSM2154806_NA_27)[c(1,2)] <- c("Gene_ID", "NA_27")
GSM2154807_NA_30 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[9]))
colnames(GSM2154807_NA_30)[c(1,2)] <- c("Gene_ID", "NA_30")
GSM2154808_NA_39 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[10]))
colnames(GSM2154808_NA_39)[c(1,2)] <- c("Gene_ID", "NA_39")
GSM2154809_NAP_3 <- fread(paste0(path_to_bulkRNA_data, bulk_file_names_list[11]))
colnames(GSM2154809_NAP_3)[c(1,2)] <- c("Gene_ID", "NAP_3")

# merge bulk data files into one table
# first, we put all the data tables into a list
bulk_RNA_Data_list <- list(GSM2154800_G1NF_42,
                           GSM2154801_GINF_44,
                           GSM2154802_GNA_73,
                           GSM2154803_GNA_78M,
                           GSM2154804_G1NF_28,
                           GSM2154805_GNA1_18,
                           GSM2154806_NA_27,
                           GSM2154807_NA_30,
                           GSM2154808_NA_39,
                           GSM2154809_NAP_3)

# Now we can use the reduce function to merge all data sets at once
# For Inner Join
bulkRNA_raw_data <- Reduce(
  function(x, y, ...) merge(x, y, ...), 
  bulk_RNA_Data_list
)

# clean up gene ID column. Remove the version ids from the gene ids
bulkRNA_raw_data$Gene_ID <- gsub("\\..*", "", bulkRNA_raw_data$Gene_ID)

# now we need to merge the GEO data with the mus Ensembl reference since the single 
# cell data uses gene names and the bulk data uses transcript_
path_to_ensembl_ref <- "C:\\Users\\hsmit\\Documents\\Saba_Lab\\R24\\Bisque_R24\\rawdata\\"
ensembl_ref <- fread(paste0(path_to_ensembl_ref, "Ensembl_Mouse_Annotation_20220107.txt"))
# remove unneeded columns from ref
ensembl_ref <- ensembl_ref %>% select("Gene stable ID",
                                      "Gene name",
                                      "Gene type")
# Merge ref data with bulk data
bulkRNA_withref_data <- merge(bulkRNA_raw_data, ensembl_ref,
                              by.x = "Gene_ID", 
                              by.y = "Gene stable ID")
# remove duplicates
bulkRNA_withref_data <- bulkRNA_withref_data %>% 
  distinct()
# make row names = gene names and remove any annotated columns
bulkRNA_Data_for_eset <- bulkRNA_withref_data[, -c(1, 13)]
bulkRNA_Data_for_eset <- bulkRNA_Data_for_eset[!duplicated(bulkRNA_Data_for_eset$`Gene name`), ]
row.names(bulkRNA_Data_for_eset) <- bulkRNA_Data_for_eset$`Gene name`
bulkRNA_rnames <- rownames(bulkRNA_Data_for_eset)
bulkRNA_Data_for_eset <- as.matrix(bulkRNA_Data_for_eset[, -c(11)])
row.names(bulkRNA_Data_for_eset) <- bulkRNA_rnames
```

# Get raw data into required eset formatting

```{r, reformat_data, include=FALSE}
##################################################
# transform bulk data into an eset data class
##################################################

# define sample names and genotype data
bulkRNA_sample_ids <- colnames(bulkRNA_Data_for_eset)
bulkRNA_subject_names <- c("Kidney_G1-APOL1_risk allele",
                           "Kidney_G1-APOL1_risk allele",
                           "Kidney_G2-APOL1_risk allele",
                           "Kidney_G2-APOL1_risk allele",
                           "Kidney_wildtype",
                           "Kidney_wildtype",
                           "Kidney_wildtype",
                           "Kidney_Control",
                           "Kidney_Control",
                           "Kidney_Control")

# build out annotation metadata
bulkRNA_pheno <- data.frame(check.names=F, check.rows=F,
                       stringsAsFactors=F,
                       row.names=bulkRNA_sample_ids,
                       SubjectName=bulkRNA_sample_ids,
                       subjectGenotype=bulkRNA_subject_names)
bulkRNA_meta <- data.frame(labelDescription=c("SubjectName",
                                         "SubjectGenotype"),
                      row.names=c("SubjectName",
                                  "SubjectGenotype"))
bulkRNA_pdata <- new("AnnotatedDataFrame",
                data=bulkRNA_pheno,
                varMetadata=bulkRNA_meta)

#build expressionSet object
bulkRNA_eset <- Biobase::ExpressionSet(assayData=bulkRNA_Data_for_eset,
                                  phenoData=bulkRNA_pdata)


# # transform single cell data into eset data class with meta data
# sample.ids <- colnames(sc.counts.matrix)
# # individual.ids and cell.types should be in the same order as in sample.ids
# sc.pheno <- data.frame(check.names=F, check.rows=F,
#                        stringsAsFactors=F,
#                        row.names=sample.ids,
#                        SubjectName=individual.labels,
#                        cellType=cell.type.labels)
# sc.meta <- data.frame(labelDescription=c("SubjectName",
#                                          "cellType"),
#                       row.names=c("SubjectName",
#                                   "cellType"))
# sc.pdata <- new("AnnotatedDataFrame",
#                 data=sc.pheno,
#                 varMetadata=sc.meta)
# sc.eset <- Biobase::ExpressionSet(assayData=bulkRNA_Data_for_eset,
#                                   phenoData=sc.pdata)
```

# Load bulk and scRNA data sets using preexisting Expression set objects from MuSiC Github repo

```{r Data_load_from_github, include=False}
path_to_eset_objects <- path_to_scRNA_data

# read in eset objects
bulkRNA_eset <- readRDS(paste0(path_to_eset_objects, "Mousebulkeset.rds"))
scRNA_eset <-  readRDS(paste0(path_to_eset_objects, "Mousesubeset.rds"))

# run bisque decomposition
res <- BisqueRNA::ReferenceBasedDecomposition(bulkRNA_eset, scRNA_eset, markers=NULL, use.overlap=FALSE)

```