---
title: "snRNA-Seq Analysis Using BNLx/SHR - Default Analysis"
author: "Harry A. Smith"
date: "`R Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    fig_caption: yes
    fig_retina: 1 
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
library(Seurat)
library(cowplot)
library(RColorBrewer)
library(Matrix)
library(tidyverse)
library(R.utils)
library(viridis)
library(patchwork)

data_dir <- "/Users/smithh/Documents/Saba_Lab/cellranger_results"
tcols <- rev(brewer.pal(11, "RdGy")[c(1:5, 7)])[1:6]
```

## Experiment Summary

Sample | Description
------------- | ---------------- 
1_BNLx_1_Neurons | BNLx Biological Replicate 1 
3_BNLx_2_Neurons | BNLx Biological Replicate 2
5_SHR_1_Neurons | SHR Biological Replicate 1  
7_SHR_2_Neurons | SHR Biological Replicate 2

```{r in_data}

samples <- c(
  "1_BNLx_1_Neurons",
  "3_BNLx_2_Neurons",
  "5_SHR_1_Neurons",
  "7_SHR_2_Neurons")

sample_paths <- file.path(data_dir ,
                          samples)

names(sample_paths) <- samples

bcells <- Seurat::Read10X(sample_paths)
```

```{r create_seurat, message = F, results = 'hide', warning = F}
bcells <- CreateSeuratObject(counts = bcells, 
                              min.cells = 3,
                              min.genes = 200, 
                              project = "Saba",
                              names.field = c(1,2), 
                              names.delim = "_")
```

```{r additional_mdata}

# add correct sample identifier
ids <- colnames(bcells@assays$RNA@data)
ids <- str_split(ids, "_[ATCG]{10}", simplify = T) %>% .[, 1]
ids <- data.frame(row.names = colnames(bcells@assays$RNA@data),
                  expt = ids)
bcells <- AddMetaData(bcells, ids)
```

  
```{r examine_umi_and_ngenes, fig.width = 10, fig.height = 8, warning = F, eval=FALSE}
summary_dat <- data_frame(cell = colnames(bcells@assays$RNA@data), 
                          expt = bcells@meta.data$expt,
                          nUMIs = colSums(as.matrix(bcells@assays$RNA@data)),
                          nGenes = colSums(as.matrix(bcells@assays$RNA@data) > 0))

summary_dat <- summary_dat %>% 
  group_by(expt) %>% 
  dplyr::arrange(desc(nUMIs))

n_cells <- summary_dat %>%  
  group_by(expt) %>%  
  summarize(n_cells = n(),
            median_UMIs = median(nUMIs),
            median_Genes = median(nGenes))

knitr::kable(n_cells,
             caption = "Number of cells prior to filtering and summary statistics")

a <- ggplot(summary_dat, aes(expt, nUMIs)) +
  geom_violin(aes(fill = expt)) +
  scale_fill_brewer(palette = "Paired") +
  ylab("Number of UMIs (Log10)") +
  theme(axis.text.x = element_blank())

b <- ggplot(summary_dat, aes(expt, nGenes)) +
    geom_violin(aes(fill = expt)) +
  scale_fill_brewer(palette = "Paired") +
  ylab("Number of Genes (at least 1 UMI) Log10") +
  theme(axis.text.x = element_blank())
plot_grid(a, b, ncol = 1)
```



```{r write_matrix, eval = F}
dir.create("/Users/smithh/Documents/Saba_Lab/cellranger_results/count_matrix", showWarnings = FALSE)

write.table(as.matrix(bcells@assays$RNA@data),
            file.path("/Users/smithh/Documents/Saba_Lab/cellranger_results/count_matrix", "all_samples_raw_counts_matrix.txt"),
            quote = F, sep = "\t")

gzip(file.path("/Users/smithh/Documents/Saba_Lab/cellranger_results/count_matrix",
               "all_samples_raw_counts_matrix.txt"),
     overwrite = TRUE,
     remove = TRUE)
```

```{r utility_fxn}
cell_counts <- function(seurat_object){
  #returns tbl_df with number of cells per sample type in seurat object
  map_df(levels(seurat_object@ident), 
         ~data_frame(sample = .x, 
                     n_cells = length(WhichCells(seurat_object, ident = .x))))
}
```

## Mitochondrial Reads {.tabset .tabset-fade}

The percentage of reads aligning to mitochondria is often an indicator of cell quality. Cells that have lysed prior to capture preferentially retain mitochondial RNA, but lose cellular cytoplasmic mRNA. As shown below there is a strong negative correlation with the number of UMIs and the proportion of reads aligning to mitochondria. Cells with less than 20% mitochondrial reads will be retained for further analysis.

```{r QC}
# there are not unique IDs for mito genes in refseq, so I  am using this gene list

mito_gene_list <- c("ATP6", "ATP8", "COX1", "COX2", "COX3", "CYTB", "ND1", "ND2", "ND3", "ND4", "ND4L", "ND5", "ND6")
mito_gene_list <- paste0(mito_gene_list, collapse = "|")

mito.genes <- grep(mito_gene_list, rownames(bcells@assays$RNA@data), value = T)
# proportion.mito <- Matrix::colSums(bcells@assays$RNA@data[mito.genes, ]) /
#   Matrix::colSums(bcells@assays$RNA@data)
# 
# bcells <- AddMetaData(bcells, proportion.mito, "proportion.mito")

bcells[["percent.mt"]] <- PercentageFeatureSet(bcells, pattern = mito_gene_list)

# unfiltered <- bcells@meta.data %>% 
#   tibble::rownames_to_column("cell") %>% 
#   left_join(summary_dat, by = c("cell", "expt")) %>% 
#   as_data_frame()
# 
# ggplot(unfiltered, aes(expt,
#                        proportion.mito)) +
#   geom_jitter(size = 0.25) +
#   geom_violin(aes(fill = expt), alpha = 0.66) +
#   scale_fill_brewer(palette = "Paired") +
#   theme(
#     axis.text.x = element_text(angle = 90, hjust = 1),
#     axis.title.x = element_blank()
#   )
```

### All Samples

```{r sample_ids}
# ggplot(unfiltered, 
#        aes(nUMI, proportion.mito)) +
#   geom_point(aes(color = expt)) +
#   scale_color_brewer(palette = "Paired")
```

```{r, results ='asis'}

sample_names <- as.character(unique(bcells@meta.data$expt))
# per_sample <- map(sample_names, ~filter(unfiltered,
                                        # expt == .x))
# 
# # generate tab with individual plot programmatically 
# # see https://stackoverflow.com/questions/43752095/programmatically-insert-header-and-plot-in-same-code-chunk-with-r-markdown-using?noredirect=1&lq=1
# 
# for(i in seq_along(per_sample)){
#   .col <- brewer.pal(10, "Paired")[i]
#   cat('\n### ', sample_names[i], '\n')
#   p <- ggplot(per_sample[[i]], aes(nUMIs, proportion.mito)) +
#         geom_point(aes(color = expt)) +
#         scale_color_manual(values = .col)
#   print(p)
#   cat('\n')
# }

# Visualize QC metrics as a violin plot

VlnPlot(bcells, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, split.by = 'expt')

```

Cells with low UMIs tend to have a higher percentage of mitochondrial reads, suggesting that these cells have lysed prematurely. However, overall a large proportion of cells are retained for further analysis after removing cells with > 20% UMIs derived from mitochondiral transcripts. 

## Filtered Data {.tabset .tabset-fade}

After filtering the distribution of UMIs, Genes, and Proportion of mitochondrial reads are shown as selectable tabs. 

```{r, apply_filters}
bcells <- subset(bcells, subset = nFeature_RNA > 250 & nFeature_RNA < 2500 & percent.mt < 5)
```

```{r ngenes_filter}
# knitr::kable(cell_counts(SetAllIdent(bcells, "expt")),
#              caption = "Number of cells passing an additional filter requiring less than 20% mitochondrial reads and at least 250 genes detected, and UMI < 2500")
```

### Post filtering violin plots
```{r ngenes_post_filter}

VlnPlot(bcells, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, 
        cols = brewer.pal(10, "Paired"),
        split.by = "expt")

```

## Preprocess the data for PCA, TSNE, and Clustering 
  The raw UMI count matrix will next be preprocessed for downstream clustering and visualization
  
### Normalize the data
  First the UMI count data will be normalized. LogNormalize: Feature counts for each cell are divided by the total counts for that cell and multiplied by the scale.factor. This is then natural-log transformed using log1p.

```{r normalize, results = 'hide'}

bcells <- NormalizeData(bcells, normalization.method = "LogNormalize", scale.factor = 10000)

#library(scran)

# sce <- SingleCellExperiment(list(counts = bcells@raw.data[rownames(bcells@data),
#                                                           rownames(bcells@meta.data)]))
# 
# qclusters <- quickCluster(counts(sce), method = "igraph", min.size = 50)
# sce <- computeSumFactors(sce, 
#                          sizes = seq(20, 100, 5),
#                          cluster = qclusters, 
#                          min.mean = 0.1)
# 
# # return normalize non-log values (log2 is default for scran, log for seurat)
# sce <- scater::normalize(sce, 
#                          return_log = F, 
#                          log_exprs_offset = 0)    
# # add pseudocount of 1 consistent with seurat
# bcells@data <- assay(sce, "normcounts") %>% 
#   log1p(.) %>% 
#   as(., "sparseMatrix")
# 
# # all seurat calc paramters are stored and occasionally checked
# # add normalization method to avoid warnings
# bcells@calc.params[["NormalizeData"]] <- list(normalization.method = "scran::computeSumFactors")

```

### Identify variable genes

  The normalized UMI count matrix contains `r  length(rownames(bcells@assays$RNA@data))` genes and `r length(colnames(bcells@assays$RNA@data))` cells. Most of these genes have very low expression or very low variance across all of the cells in the dataset. These genes are therefore very uninformative for classifying different cell types. In this next step the set of variable genes useful for classification will be identified, and used for downstream PCA, TSNE, and classification purposes. 
  
  Variable genes will be selected by examining a plot of the average expression across all cells on the x axis to the dispersion across all genes. Highly variable genes will have non-zero dispersion. The dispersion is defined by the variance / mean. 
  
```{r get_variable_genes, results = 'hide'}
# bcells <- FindVariableFeatures(bcells, 
#                     mean.function = ExpMean, 
#                     dispersion.function = LogVMR, 
#                     x.low.cutoff = 0.1, 
#                     x.high.cutoff = 5, 
#                     y.cutoff = 0.4, do.contour = F, do.plot = T)

bcells <- FindVariableFeatures(
  bcells,
  selection.method = "vst",
  verbose = TRUE)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(bcells), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(bcells)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
plot2
```

### Control for noise
  Single-cell data is often noisy, and it is difficult to compare cells with very high UMIs to very low UMIs. For classification we do not want to bias our clusters based on nuisance variables such as UMI count or proportion of mitochondria. To help mitigate bias in our clustering a regression model is fit to predict gene expression based on nuisance variables. The residuals from this fit are then scaled and centered (i.e. turned into z-scores) for classification purposes. The scaled regressed data is not used for differential expression, but used only for classifcation, TSNE visualization, and pseudotime analysis. 
  
From function description: Scales and centers features in the dataset. If variables are provided in vars.to.regress, they are individually regressed against each feature, and the resulting residuals are then scaled and centered.

```{r scale_data, results='hide', message = F}
all.genes <- rownames(bcells)
bcells <- ScaleData(bcells, features = all.genes, split.by = 'expt')
```

### Calculate Principal Components

  PCA will now be performed on processed data using only the set of variable genes defined above. PCA will generate a set of new vectors that describe the variablity in the data. A subset of these principal components will be selected to be used as input data for TSNE visualization. Selecting the proper number of components is informed by examining the amount of variance captured in each PC. Shown below is a plot of variance captured in each PC. At a certain point, additional PCs capture little additional varaince, and therefore are uninformative for classification. For this dataset 15 PCs will be used for classification.
  
```{r PCA_calc, results='hide'}
bcells <- RunPCA(bcells, features = VariableFeatures(object = bcells))

VizDimLoadings(bcells, dims = 1:2, reduction = "pca")

DimPlot(bcells, reduction = "pca")
DimPlot(bcells, reduction = "pca", split.by = 'expt')

# PCElbowPlot(bcells)
```

### PCA Heatmaps {.tabset .tabsetfade}
  
  
```{r pc_plots, results ='asis'}
DimHeatmap(bcells, dims = 1:15, cells = 500, balanced = TRUE)


# for(i in seq_along(pcs)){
#   cat('\n#### ', 'PC', pcs[[i]][1], ' vs PC', pcs[[i]][2], '\n', sep = "")
#   DimPlot(bcells, 
#               dims =  c(pcs[[i]][1], 
#                pcs[[i]][2]),
#                col = brewer.pal(10, "Paired"), 
#         group.by = "expt")
#   #print(p)
#   cat('\n')
# }
```

### Determining dimensionality of the data {.tabset .tabset-fade}

```{r, dim_determine, results ='asis'}
bcells <- JackStraw(bcells, num.replicate = 100)
bcells <- ScoreJackStraw(bcells, dims = 1:20)
```

#### JackStraw Plot

```{r js_plot, results ='asis'}
JackStrawPlot(bcells, dims = 1:15)
```

#### Elbow Plot

```{r elbow_plot, results ='asis'}
ElbowPlot(bcells)
```

### Cluster cells {.tabset .tabsetfade}
  
    
```{r select_dims}
bcells <- FindNeighbors(bcells, dims = 1:15)
bcells <- FindClusters(bcells, resolution = 0.5)

# reticulate::py_install(packages =
#  'umap-learn')

bcells <- RunUMAP(bcells, dims = 1:15)

DimPlot(bcells, reduction = "umap")

DimPlot(bcells, reduction = "umap", split.by = 'expt')

# save Seurat Object
# saveRDS(bcells, file = paste0(data_dir, "/", "bcells_default.rds"))
```



## Differential Expression and Marker Gene Identification

```{r find_markers, eval=FALSE}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
bcells.markers <- FindAllMarkers(bcells, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
bcells.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

write_csv(bcells.markers, file = paste0(data_dir, "/", "all_markers_all_clusters.csv"))
```


## Session Info
```{r ses}
sessionInfo()
```